src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -508,5 +512,57 @@ public static void EndNoGCRegion()
         {
             EndNoGCRegionWorker();
         }
+
+        // generation: -1 means to let GC decide (equivalent to regular new T[])
+        // 0 means to allocate in gen0
+        // GC.MaxGeneration means to allocate in the oldest generation
+        //
+        // pinned: true means you want to pin this object that you are allocating
+        // otherwise it's not pinned.
+        //
+        // alignment: only supported if pinned is true.
+        // -1 means you don't care which means it'll use the default alignment.
+        // otherwise specify a power of 2 value that's >= pointer size
+        // the beginning of the payload of the object (&array[0]) will be aligned with this alignment.
+        // static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        public static T[] AllocateArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        {
+            // TODO: VS resources
+            if (generation != -1)
+                throw new NotSupportedException();
+
+            if (pinned)
+                throw new NotSupportedException();
+
+            if (alignment != -1)
+                throw new NotSupportedException();
+
+            return AllocateArrayWorker<T>(length, clearMemory: true);
+        }
+
+        // Skips zero-initialization of the array if possible. If T contains object references, 
+        // the array is always zero-initialized.
+        public static T[] AllocateUninitializedArray<T>(int length, int generation = -1, bool pinned = false, int alignment = -1)
+        {
+            // TODO: VS resources
+            if (generation != -1)
+                throw new NotSupportedException();
+
+            if (pinned)
+                throw new NotSupportedException();
+
+            if (alignment != -1)
+                throw new NotSupportedException();
+
+            return AllocateArrayWorker<T>(length, clearMemory: false);
+        }
+
+        private static T[] AllocateArrayWorker<T>(int length, bool clearMemory)
```

Comment:
It is not profitable to have a helper like this for generics (the generic instantations are not free, and shared generics inline poorly). Duplicate the check instead.